SwrveSDK/src/main/java/com/swrve/sdk/messaging/view/SwrveMessageView.java

```diff
@@ -1,251 +1,312 @@
+                    } else {
+                        child.layout(st.leftMargin - cCenterX + centerx, st.topMargin - cCenterY + centery, st.leftMargin + cCenterX + centerx, st.topMargin + cCenterY + centery);
                     }
-
-                    @Override
-                    public void onAnimationRepeat(Animation anim) {
-                    }
-
-                    @Override
-                    public void onAnimationEnd(Animation anim) {
-                        dismissView();
-                    }
-                });
-                startAnimation(animation);
-            } else {
-                dismissView();
+                }
             }
         } catch (Exception e) {
-            SwrveLogger.e(LOG_TAG, "Error while dismissing message", e);
+            SwrveLogger.e(LOG_TAG, "Error while onLayout in SwrveMessageView", e);
         }
     }
 
-    private void dismissView() {
-        final ViewParent parent = getParent();
-        if (containerDialog == null) {
-            removeView(parent);
-        } else {
-            containerDialog.dismiss();
-        }
+    protected SwrveImageView createSwrveImage(Context context) {
+        return new SwrveImageView(context);
     }
 
-    protected void removeView(ViewParent parent) {
-        if (parent != null) {
-            ((ViewGroup) parent).removeView(SwrveMessageView.this);
-        }
-        destroy();
+    protected SwrveButtonView createSwrveButton(Context context, SwrveActionType type) {
+        return new SwrveButtonView(context, type);
     }
 
     @Override
     public boolean onTouchEvent(MotionEvent event) {
         return true;
     }
 
-    @Override
-    public void draw(Canvas canvas) {
-        super.draw(canvas);
-
+    public void destroy() {
         try {
-            if (firstTime && firstDraw) {
-                firstDraw = false;
-                message.getCampaignManager().messageWasShownToUser(format);
+            if (bitmapCache != null) {
+                // Iterate through all the bitmaps to recycle them
+                Iterator<WeakReference<Bitmap>> bitmapIt = bitmapCache.iterator();
+                while (bitmapIt.hasNext()) {
+                    WeakReference<Bitmap> weakBitmap = bitmapIt.next();
+                    Bitmap b = weakBitmap.get();
+                    if (b != null) {
+                        if (!b.isRecycled()) {
+                            b.recycle();
+                        }
+                        b = null;
+                    }
+                    weakBitmap = null;
+                }
+
+                bitmapCache.clear();
+                bitmapCache = null;
             }
-        } catch (Exception e) {
-            SwrveLogger.e(LOG_TAG, "Error while processing first impression", e);
+            System.gc();
```

Comment:
Force liberate bitmaps. Depending on the API they are not tied to the Java object and might not be GCed in time for another IAM to show and boom, OutOfMemoryException